Improve call performance for `sync_partition_metadata` utility

fgwang7w opened the pull request:
**This PR fixes following issues:**

> Case 1: Current implementation for getting partition names takes partition values directly without taking into account of the actual path. e.g. The use of escaped character maybe encoded in HMS(Hive MetaStore) as partition name that could result incorrectness with current naming convention to fetch partition names.

- Solution: Retrieve partition storage location correctly from metastore using correct API calls

> Case 2: Sync partition utility lists file status of each partition twice. In case of handling a large of number of partitions for an external table, redundant check file status is a performance bottleneck.
- Solution: Avoid file status check after knowing the path name from storage already by the sync partition procedure code path.

> Case 3: For create partition call, batch size for number of trunks of partitions to be added to the metastore database is bounded by a fixed number. This limits the performance to update the metastore when the consumer of the underlying database is capable to handle a large number of batch and a large size of update batch. 
- Solution: Different metastore systems could implement this commit batch size differently based on different underlying database capacity.

<br>

== RELEASE NOTES ==

```
General Changes
* improve performance for procedure sync_partition_metadata 

Hive Changes
* Increase default partition batch size per commit for Hive Metastore up to 10. 
* Set default partition batch size per commit for Glue Metastore up to 10,000
```

Detail troubleshoot and solution design can be found in this reference page [here](https://docs.google.com/document/d/1BL9Qtrma6YJcPT3sZc3ofMgYDUWRz0kTKFYEJL0aMwo/edit?usp=sharing)


| # of partitions | Before (min) | after(min) | delta|
| :---: | :---: | :---: | :---: |
| 2,526 | 5 |  1 |  5 |
| 12,000 | 54 | 6 | 9 |
| 39,493 | 92 | 12 | 7 |

agrawalreetika reviewed (changes requested):
Thank you for working on this.
The Benchmark number looks promising, this would be a good improvement for the sync-partition procedure.
I have taken the first pass on PR and added my comments.
These attached Benchmark results are from glue or HMS? Should we do it for both?

fgwang7w commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -114,10 +114,9 @@
     private static final int MAX_LAST_DATA_COMMIT_TIME_ENTRY_PER_TABLE = 100;
     private static final int MAX_LAST_DATA_COMMIT_TIME_ENTRY_PER_TRANSACTION = 10_000;
-
+    private static final int PARTITION_COMMIT_BATCH_SIZE = 3000;
```
Initially the high bound was set for performance testing but it could end up sending all partitionadd info in a single thrift call which could potentially cause metastore timeout error. So it's lowered to 3000 for now for sake of improving how memory is used in the metastore and to avoid client read timeout exceptions.

I believe for glue catalog it's safe with high bound since `BatchCreatePartitionRequest` is sent to glue metastore with 100 partition in a batch. I think the concern is for HMS and other metastores. 

I am open to make a reasonable change to define the value of this threshold.

agrawalreetika replied:
Yeah, for Glue it's capping it to 100. My only concern is about HMS thinking 3000 could be an overwhelming number for maybe smaller HMS instances

fgwang7w replied:
ok it makes sense to align the limit to 100

fgwang7w commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -127,8 +133,17 @@ private void doSyncPartitionMetadata(ConnectorSession session, String schemaName
+            List<String> partitionNamesInMetastore = metastore.getPartitionNames(new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), getMetastoreHeaders(session), isUserDefinedTypeEncodingEnabled(session), metastore.getColumnConverterProvider()), schemaName, tableName)
                     .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            List<String> partitionsInMetastore = new ArrayList<>(partitionNamesInMetastore.size());
+            for (List<String> batchPartitionNames : Lists.partition(partitionNamesInMetastore, BATCH_GET_PARTITION_BY_NAMES)) {
```
Yes Glue metastore implementation supports `getPartitionByNames` in batch with an upper limit of 1000 partitions per request to aws. So i think it should be safe to remove the code to break partitions in batch in sync partition phase.

imjalpreet commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -127,8 +132,17 @@ private void doSyncPartitionMetadata(ConnectorSession session, String schemaName
                     .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            List<String> partitionsInMetastore = new ArrayList<>(partitionNamesInMetastore.size());
+
+            Map<String, Optional<Partition>> partitionsOptionalMap = metastore.getPartitionsByNames(metastoreContext, schemaName, tableName, partitionNamesInMetastore);
```
I had a quick question. Since we just need the list of partition names, why are adding this additional method call to fetch partitions and then creating a new list of partition names? Wouldn't the lists `partitionNamesInMetastore` and `partitionsInMetastore` have the same values?

Is there a mismatch in the output of `metastore.getPartitionNames` and `metastore.getPartitionsByNames`?

fgwang7w replied:
I actually put down the reasoning in the commit message:

``Current implementation for getting partition names is to take partition values directly
without take into account of the actual path name.
According to the hive source code, this main logic in hive may change that will impact the
 validity of the partition names exists in the storage.
https://github.com/apache/hive/blob/6ada437afc641cd44edf4ea0ccdd905a37385005/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/FileUtils.java#L246
``

imjalpreet replied:
Okay, I understand. But we are still not making use of the actual path, in this case, we still just added the values directly using entry.getKey() which will be the same value as what we get from metastore.getPartitionNames. If we want to achieve what the above link suggests, shouldn't we make use of the partition path to generate this value?

fgwang7w replied:
fixed, good catch, thx

imjalpreet replied:
Also, I believe we should convert this method call into multiple batches. GlueHiveMetastore implementation already does it but ThriftHiveMetastore will fetch all partitions at once which will result in timeouts or very slow fetches in tables with a considerable number of partitions.

agrawalreetika replied:
+1
Even I noticed now batches are being made only for GlueHiveMetastore implementation but not for other metastores. We can have this call the way we had it earlier or we can extend batch-level implementation even for other megastores like it's being done for Glue?

fgwang7w replied:
agree, the caller api needs to guard the load not overwhelm the underlying metastore.

imjalpreet commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -61,6 +64,7 @@
 public class SyncPartitionMetadataProcedure
         implements Provider<Procedure>
 {
+    private static final int BATCH_GET_PARTITION_BY_NAMES = 1000;
```
nit: Unused variable

imjalpreet reviewed (changes requested):
@fgwang7w Thank you for the optimisation. I did the first pass and I have a question regarding one change added in SyncPartitionMetadataProcedure.

imjalpreet reviewed (changes requested):
I have one more suggestion, please let me know what are your views on the same.

agrawalreetika reviewed (approved):
The changes look good to me. Quick question, Are these attached Benchmark results from glue? Could we update the document with the same?

fgwang7w commented:
> here
sure, will update the release notes. many thanks for help to review.

fgwang7w commented:
> The changes look good to me. Quick question, Are these attached Benchmark results from glue? Could we update the document with the same? Please update the PR release notes too about the partition batch.

sure, will update the release notes. many thanks for help to review.

yzhang1991 commented:
@tdcmeehan @prestodb/committers Hi Presto committers, this PR is reviewed internally already, can we have a second round review?

pranjalssh commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -841,31 +843,36 @@ private List<Partition> batchGetPartition(String databaseName, String tableName,
     @Override
     public MetastoreOperationResult addPartitions(MetastoreContext metastoreContext, String databaseName, String tableName, List<PartitionWithStatistics> partitions)
     {
+        List<CompletableFuture<Future<BatchCreatePartitionResult>>> futures = new ArrayList<>();
```
This can be just List<CompletableFuture<BatchCreatePartitionResult>>

fgwang7w replied:
this is used to list all CompletableFuture to hold each async batch request for which each batch request is a future call.

pranjalssh replied:
Type of this is what is returned in thenApply call(which is just null). So you can make this List<CompletableFuture<Void>>

pranjalssh commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -841,31 +843,36 @@ private List<Partition> batchGetPartition(String databaseName, String tableName,
     @Override
     public MetastoreOperationResult addPartitions(MetastoreContext metastoreContext, String databaseName, String tableName, List<PartitionWithStatistics> partitions)
     {
+        List<CompletableFuture<Future<BatchCreatePartitionResult>>> futures = new ArrayList<>();
```
What is the performance benefit of changing this function? Can you run benchmarking with and without this optimization?

Earlier, the code as doing these in parallel as well. Once you create a list of futures, they all start running in parallel. So there should not be any difference if you await on them in serial or parallel as we want all to finish.

fgwang7w replied:
I add more comments near code blocks to explain, but the idea is to create partition request via BatchCreatePartitionResult as a CompletableFuture, so it won't be a blocking call. The previous version of logic does batch-by-batch fashion instead of all-by-once fashion.

pranjalssh replied:
Previous version isn't batch-by-batch. Futures are created by executorService.submit(), so computation has started in background as soon as you call batchCreatePartitionAsync().

fgwang7w replied:
yes the computation is kicked off via async request but is it true that previous version at the end wait and check for the callback one-by-one? It makes more sense to optimize this code using `CompletableFuture` interface and wait for callbacks to complete at the end via `CompletableFuture.allOf`.

pranjalssh replied:
@fgwang7w yes it checks results one by one, but it wouldn't make a difference as calling get() one by one is same as using a list of completable futures. This smells like micro-optimization. Can you run the benchmark without this?

fgwang7w replied:
Just tested sync partition code path against 39k partitions with and without this code block change, and the result is very close. Per offline discussion, I will rollback this code block change to make the code more readable instead of using CompletableFuture

arunthirupathi commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -115,10 +115,9 @@
     private static final int MAX_LAST_DATA_COMMIT_TIME_ENTRY_PER_TABLE = 100;
     private static final int MAX_LAST_DATA_COMMIT_TIME_ENTRY_PER_TRANSACTION = 10_000;
-
+    private static final int PARTITION_COMMIT_BATCH_SIZE = 100;
```
What is the impact of this on other metastores ? would it cause the request to spike and cause cascading issues ?

fgwang7w replied:
Current Hive versions of metastore backend with RDBMS should be able to handle 10000+ partitions.. Glue has an upper bound on 100 partitions per batch. information is listed in this commit message.

fgwang7w replied:
sure will add comment to note the batch size setting

arunthirupathi replied:
SemiTransactionalHiveMetastore is used for all Metastores including Meta's metastore and Hudi and other metastores. Increasing the limit from 8 to 100 is not that simple and might over whelm some of the metastores. 

There are more metastores than Glue and old HiveMetastores.

arunthirupathi replied:
This should at least go behind a config and default should be left to 8, or more careful analysis is required.

arunthirupathi replied:
Better solution is to change the API, so that the final callers can decide how much they want to batch instead of batching upstream.

fgwang7w replied:
good point, given there are various underlying metastore systems, let's give a default value of 10 in general, and caller will decide the batch partition size for commit, e.g. glue will have 100 partitions per guarded by `BATCH_CREATE_PARTITION_MAX_PAGE_SIZE` and process up to 100 requests in parallel so that BATCH_PARTITION_COMMIT_TOTAL_SIZE is 10000.

fgwang7w commented:
hello @arunthirupathi @pranjalssh @agrawalreetika @imjalpreet  thank you for helping reviewing this PR, I have addressed all issues, is there any other concerns or suggestions? @yingsu00 thank you for giving a review in advance.

pranjalssh commented:
@rschlussel for merge

yingsu00 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -823,6 +823,20 @@ public synchronized void addPartition(
+            Partition partition,
+            Path currentLocation,
+            PartitionStatistics statistics,
+            boolean partitionActionBySync)
```
Rename to `needToValidatePath`

yingsu00 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -2409,14 +2427,26 @@ public String toString()
         private final Optional<List<String>> fileNames;
         private final PartitionStatistics statistics;
         private final PartitionStatistics statisticsUpdate;
+        private final boolean partitionActionBySync;
```
It seems all members in PartitionAndMore are objects/nouns that describes a partition. But partitionActionBySync is a source of action. Is there another better way we can pass it in to prepareAddPartition?

fgwang7w replied:
I thought about optimizing the implementation but `prepareAddPartition` is at commit phase which is too late to be aware of the state of added partition whether it's coming from the general insert path or sync utility logic.

fgwang7w replied:
Thank you @yingsu00 for suggestion. It makes sense not to add a source of action in `PartitionAndMore`, the flag instead is now added to `HDFSContext` structure to indicate whether path validation is needed or not.

yingsu00 reviewed (commented):
@fgwang7w George, great improvement! In addition to the comments in code, there are several other nits:
1. Commit title for 1 and 3 are too long
2. Commit title first letter shall be upper case
3. Can you use full sentences in the PR message?

fgwang7w commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -28,6 +28,7 @@
 import com.facebook.presto.spi.procedure.Procedure.Argument;
```
makes sense, thanks

fgwang7w commented:
Thank you @yingsu00 for review. All comments are addressed, could you please give another pass? Thanks!
Thank you @pranjalssh @agrawalreetika for approving the PR.

fgwang7w commented:
@prestodb/committers ping!

yingsu00 commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -127,8 +141,17 @@ private void doSyncPartitionMetadata(ConnectorSession session, String schemaName
-            List<String> partitionsInMetastore = metastore.getPartitionNames(new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), getMetastoreHeaders(session), isUserDefinedTypeEncodingEnabled(session), metastore.getColumnConverterProvider()), schemaName, tableName)
+            List<String> partitionNamesInMetastore = metastore.getPartitionNames(metastoreContext, schemaName, tableName)
                     .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            List<String> partitionsInMetastore = new ArrayList<>(partitionNamesInMetastore.size());
```
Use ImmutableList and ImmutableList.builder().builder()

fgwang7w replied:
fixed

yingsu00 commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -61,6 +65,8 @@
 public class SyncPartitionMetadataProcedure
         implements Provider<Procedure>
 {
+    private static final int BATCH_GET_PARTITION_BY_NAMES = 1000;
```
GET_PARTITION_BY_NAMES_BATCH_SIZE

yingsu00 commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -28,6 +28,7 @@
 import com.facebook.presto.spi.procedure.Procedure.Argument;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
```
Commit title too long. Upper case the first Letter.

fgwang7w replied:
fixed

yingsu00 commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HdfsContext.java:
```diff
@@ -32,6 +32,7 @@
     private final Optional<String> tablePath;
     // true if the table already exist in the metastore, false if the table is about to be created in the current transaction
     private final Optional<Boolean> isNewTable;
+    private final Optional<Boolean> noNeedToValidatePath;
```
noNeedToValidatePath -> isPathValidationNeeded

fgwang7w replied:
ok revised

yingsu00 commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HdfsContext.java:
```diff
@@ -92,6 +94,22 @@ public HdfsContext(ConnectorSession session, String schemaName, String tableName
+            String tableName,
+            String tablePath,
+            boolean isNewTable,
+            boolean noNeedToValidatePath)
```
noNeedToValidatePath -> isPathValidationNeeded

yingsu00 commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HdfsContext.java:
```diff
@@ -92,6 +94,22 @@ public HdfsContext(ConnectorSession session, String schemaName, String tableName
+                Optional.of(requireNonNull(tableName, "tableName is null")),
+                Optional.of(requireNonNull(tablePath, "tablePath is null")),
+                Optional.of(isNewTable),
+                Optional.of(noNeedToValidatePath));
```
noNeedToValidatePath -> isPathValidationNeeded Same for all other occurrences

yingsu00 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -1488,19 +1515,23 @@ private void prepareAddPartition(MetastoreContext metastoreContext, HdfsContext
-                            targetPath,
-                            () -> cleanUpTasksForAbort.add(new DirectoryCleanUpTask(context, targetPath, true)));
+            // we can bypass the file storage path checking logic for sync partition code path
+            // because partition path locality has been determined already during early phase of sync logic to identify the partition information
```
// because the file paths have been verified during early phase of the sync logic already

yingsu00 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -150,7 +150,29 @@
     private static final String DEFAULT_METASTORE_USER = "presto";
     private static final String WILDCARD_EXPRESSION = "";
+
+    /** This is the total number of partitions allowed to process in a big batch chunk which splits multiple smaller batch of partitions allowed by BATCH_CREATE_PARTITION_MAX_PAGE_SIZE
```
No need to use the document comment /**...*/ which would be picked up by the HTML extractors. This is a private member field.

yingsu00 reviewed (commented):
@fgwang7w Mostly good, just some nits.
In the PR message and commit messages:
> without take into account

-> without taking into account

> hms 

-> HMS(Hive Meta Store)

> Case 2: Sync partition utility lists file status of each partition twice. In case of handling a large of number of partitions for an external table, redundant check file status is a performance bottleneck.

Case 2: Sync partition utility lists file status of each partition twice, causing performance bottleneck when processing an external table with a large of number of partitions.

Commit 1 & 3 titles are too long. 
Use upper case for the first letter in each sentence, including all commit titles, commit/PR messages, newly added comments.

fgwang7w commented on presto-hive/src/main/java/com/facebook/presto/hive/SyncPartitionMetadataProcedure.java:
```diff
@@ -127,8 +141,17 @@ private void doSyncPartitionMetadata(ConnectorSession session, String schemaName
         try {
             FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, tableLocation);
-            List<String> partitionsInMetastore = metastore.getPartitionNames(new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), getMetastoreHeaders(session), isUserDefinedTypeEncodingEnabled(session), metastore.getColumnConverterProvider()), schemaName, tableName)
+            List<String> partitionNamesInMetastore = metastore.getPartitionNames(metastoreContext, schemaName, tableName)
```
changing to `partitionNames` as a general term may be confusing because there are many other instance variables that hold account for partition names, I'd propose to keep current naming for this variable explicitly defined
